Refactor: multi process to multi thread in pocket

larcane97 opened the pull request:
## Desc
- edit multi process to multi thread in pocket

## Task

- [x] change multi process to multi thread in server
- [x] remove unused vars in pocket_server
    ```python
    pipe: mp.Pipe
    process: mp.Process
    future_store: dict[str, asyncio.Future]
    torn_down: bool = False
    _uidset: set
    _cores: dict[str, PocketCore]
    ```
- [x] add new vars in pocket_server
    ```python
    thread: threading.Thread
    _initialized: bool
    _initialization_event: threading.Event
    _initialization_error: Optional[Exception]
    ```
- [x] only check the `PocketMain` instance count to keep pocket server running
- [x] edit pocket main method name
  - same to pocket core
- [x] add some future resolve logic in `FutureStore`
  - check loop is running
- [x] change some methods to async def in `PocketCore` and `PocketAuth`
- [x] add wait logic in `get_session_state` in `PocketAuth`
- [x] move registering hyperdock_wasm endpoint logic into `PocketServer`
  - and remove api router interface in dock.

Commit:
refactor: multiprocess to multithread

Commit:
edit pocket main method name

Commit:
edit langchain example

Commit:
set loop as class var in futureStore

Commit:
move loop var to futuredata

Commit:
remove loop var in futurestore

Commit:
change sync def to async def in pocket_core and pocket_auth

Commit:
edit examples

Suckzoo commented on libs/hyperpocket/hyperpocket/server/server.py:
```diff
@@ -1,101 +1,91 @@
-            dock_routes.update(dock_route)
-            self.fastapi_app.include_router(dock.router)
+            while self.thread.is_alive():
+                await asyncio.sleep(0.1)
```
Why `asyncio.sleep(0.1)` rather than `asyncio.sleep(0)` or `yield`? Is there any specific reason?

Suckzoo replied:
Also, removing dock router integration breaks WASM dock behavior.

larcane97 replied:
no i had no any specific reason.
so i will change this to `asyncio.sleep(0)`.

thank you for pointing it out !!

Suckzoo commented on libs/hyperpocket/hyperpocket/pocket_main.py:
```diff
@@ -1,18 +1,18 @@
     core: PocketCore
-    _uid: str
+
+    _cnt_pocket_count: int = 0
```
Also for ownership of this variable: why did you moved ownership of this variable? i.e. why pocket decides server destruction rather than deciding in server side?

larcane97 replied:
thank you for pointing that out.
i will change it as thread-safe integer soon.

and at first, i think because `PocketMain` launched `PocketServer`, so `PocketServer`'s owner is `PocketMain`.

also, the `PocketServer` is a singletone, whereas `PocketMain` isn't.
so if the `PocketServer` is responsible for counting and managing `PocketCore`, the logic could become more complicated, i think.

Suckzoo commented on libs/hyperpocket/hyperpocket/pocket_auth.py:
```diff
@@ -112,19 +112,21 @@ def check(
             return AuthState.NO_SESSION
 
         if session.auth_resolve_uid:
+            await asyncio.sleep(0)
```
Why this function has to be a coroutine?

larcane97 replied:
i wanted to wait just before checking the future, because some future resolving process is running in other thread.
so, i had to add asyncio.sleep, and i made this method as async function.

Suckzoo replied:
If that is the intention, yielding seems not enough nor necessary.

larcane97 replied:
can you explain it more? or do you have any suggestion about it?
and plus, i added a comment on the code.

Suckzoo commented on libs/hyperpocket/hyperpocket/futures/futurestore.py:
```diff
@@ -42,7 +41,13 @@ def resolve_future(self, uid: str, value: Any):
+            # if the future loop is running, it should be executed in same event loop
+            loop = future_data.future.get_loop()
+            if loop.is_running():
+                loop.call_soon_threadsafe(future_data.future.set_result, value)
```
👍

Commit:
add hyperdock-wasm endpoint

larcane97 commented on libs/hyperpocket/hyperpocket/pocket_main.py:
```diff
@@ -319,42 +294,26 @@ async def tool_call_in_subprocess(
     def _teardown_server(self):
-        self.server.teardown()
-    
+        self.teardown()
```
@Suckzoo 
i left this method for backward compatibility.
this method is currently used in some test code.

once all refactoring is done, i will fix all ci tests.
at the time, i will remove this method.

Commit:
make _cnt_pocket_count to atomic integer

Commit:
change sleep time from 0.1 to 0 in server teardown

Commit:
add hyperdock-wasm endpoint

Commit:
Merge branch 'main' into moon/refactor/go-to-runtime-and-dock

Commit:
refactor: runtime/dock, tool_ref, tool_request
- remove tool_request
- remove tool_ref
- mapping dock and tool 1 on 1

- TODO
  - wasm browser
  - test code

Commit:
remove tool_ref and repository dir

Commit:
edit default git source path
- remove encoded path
- edit container tag name

Commit:
remove hyperdock-wasm and roll back runtime into hyperdock

Commit:
set function tool description default as tool name

Commit:
change dock image_tag
- key: request_path + tool_name

Commit:
remove pocket core

Commit:
remove pocket core

Commit:
edit git cache and docker image tag to use git_sha

Commit:
add comment

Commit:
add image_postfix in dock
- tool_path + git_sha

Commit:
remove unnecessary asyncio.sleep(0)

larcane97 committed:
Merge pull request #114 from vessl-ai/moon/refactor/go-to-runtime-and-dock

refactor: runtime/dock
